File: src/config/providers.ts
export const DEFAULT_OPENAI_BASE_URL = 'https://api.openai.com/v1'
export const DEFAULT_OPENROUTER_OPENAI_BASE_URL = 'https://openrouter.ai/api/v1'
export const DEFAULT_CANOPYWAVE_OPENAI_BASE_URL = 'https://inference.canopywave.io/v1'
export const DEFAULT_GEMINI_OPENAI_BASE_URL =
  'https://generativelanguage.googleapis.com/v1beta/openai'
export const DEFAULT_ANTHROPIC_OPENAI_BASE_URL = 'https://api.anthropic.com/v1'
export const DEFAULT_GROK_OPENAI_BASE_URL = 'https://api.x.ai/v1'
export const OPENCODEGO_DEFAULT_BASE_URL = 'https://opencode.ai/zen/go/v1'
type ReasoningEffort = 'low' | 'medium' | 'high'

export type ProviderTransport = 'chat_completions'

export type ResolvedProviderRequest = {
  transport: ProviderTransport
  requestedModel: string
  resolvedModel: string
  baseUrl: string
  reasoning?: {
    effort: ReasoningEffort
  }
}

type ModelDescriptor = {
  raw: string
  baseModel: string
  reasoning?: {
    effort: ReasoningEffort
  }
}

const LOCALHOST_HOSTNAMES = new Set(['localhost', '127.0.0.1', '::1'])

function asTrimmedString(value: unknown): string | undefined {
  return typeof value === 'string' && value.trim() ? value.trim() : undefined
}

function parseReasoningEffort(value: string | undefined): ReasoningEffort | undefined {
  if (!value) return undefined
  const normalized = value.trim().toLowerCase()
  if (normalized === 'low' || normalized === 'medium' || normalized === 'high') {
    return normalized
  }
  return undefined
}

function parseModelDescriptor(model: string): ModelDescriptor {
  const trimmed = model.trim()
  const queryIndex = trimmed.indexOf('?')
  if (queryIndex === -1) {
    return {
      raw: trimmed,
      baseModel: trimmed,
    }
  }

  const baseModel = trimmed.slice(0, queryIndex).trim()
  const params = new URLSearchParams(trimmed.slice(queryIndex + 1))
  const reasoning = parseReasoningEffort(params.get('reasoning') ?? undefined)

  return {
    raw: trimmed,
    baseModel,
    reasoning: typeof reasoning === 'string' ? { effort: reasoning } : reasoning,
  }
}

export function isLocalProviderUrl(baseUrl: string | undefined): boolean {
  if (!baseUrl) return false
  try {
    return LOCALHOST_HOSTNAMES.has(new URL(baseUrl).hostname)
  } catch {
    return false
  }
}

export function resolveProviderRequest(options?: {
  model?: string
  baseUrl?: string
  fallbackModel?: string
}): ResolvedProviderRequest {
  const requestedModel =
    options?.model?.trim() ||
    process.env.OPENAI_MODEL?.trim() ||
    options?.fallbackModel?.trim() ||
    'gpt-4o'
  const descriptor = parseModelDescriptor(requestedModel)
  const rawBaseUrl =
    options?.baseUrl ??
    process.env.OPENAI_BASE_URL ??
    process.env.OPENAI_API_BASE ??
    undefined
  const transport: ProviderTransport = 'chat_completions'

  return {
    transport,
    requestedModel,
    resolvedModel: descriptor.baseModel,
    baseUrl: (rawBaseUrl ?? DEFAULT_OPENAI_BASE_URL).replace(/\/+$/, ''),
    reasoning: descriptor.reasoning,
  }
}

File: config/providers.go
package config

import (
	"net/url"
	"os"
	"strings"
)

// Default base URLs for each provider.
const (
	DefaultOpenAIBaseURL           = "https://api.openai.com/v1"
	DefaultOpenRouterOpenAIBaseURL = "https://openrouter.ai/api/v1"
	DefaultCanopyWaveOpenAIBaseURL = "https://inference.canopywave.io/v1"
	DefaultGeminiOpenAIBaseURL     = "https://generativelanguage.googleapis.com/v1beta/openai"
	DefaultAnthropicOpenAIBaseURL  = "https://api.anthropic.com/v1"
	DefaultGrokOpenAIBaseURL       = "https://api.x.ai/v1"
	DefaultOpenCodeGoBaseURL       = "https://opencode.ai/zen/go/v1"
)

// ProviderTransport is the transport type for provider requests.
type ProviderTransport string

const TransportChatCompletions ProviderTransport = "chat_completions"

// ReasoningEffort levels.
type ReasoningEffort string

const (
	ReasoningLow    ReasoningEffort = "low"
	ReasoningMedium ReasoningEffort = "medium"
	ReasoningHigh   ReasoningEffort = "high"
)

// ResolvedProviderRequest holds the resolved provider request details.
type ResolvedProviderRequest struct {
	Transport      ProviderTransport `json:"transport"`
	RequestedModel string            `json:"requested_model"`
	ResolvedModel  string            `json:"resolved_model"`
	BaseURL        string            `json:"base_url"`
	Reasoning      *struct {
		Effort ReasoningEffort `json:"effort"`
	} `json:"reasoning,omitempty"`
}

// IsLocalProviderURL checks if a base URL points to localhost.
func IsLocalProviderURL(baseURL string) bool {
	if baseURL == "" {
		return false
	}
	u, err := url.Parse(baseURL)
	if err != nil {
		return false
	}
	h := u.Hostname()
	return h == "localhost" || h == "127.0.0.1" || h == "::1"
}

// ResolveProviderRequest resolves model/baseUrl/transport from options and env.
func ResolveProviderRequest(model, baseURL, fallbackModel string) ResolvedProviderRequest {
	requested := strings.TrimSpace(model)
	if requested == "" {
		requested = strings.TrimSpace(os.Getenv("OPENAI_MODEL"))
	}
	if requested == "" {
		requested = strings.TrimSpace(fallbackModel)
	}
	if requested == "" {
		requested = "gpt-4o"
	}

	resolvedModel := requested
	var reasoning *struct {
		Effort ReasoningEffort `json:"effort"`
	}
	if idx := strings.Index(requested, "?"); idx != -1 {
		resolvedModel = strings.TrimSpace(requested[:idx])
		params, _ := url.ParseQuery(requested[idx+1:])
		if r := strings.TrimSpace(strings.ToLower(params.Get("reasoning"))); r == "low" || r == "medium" || r == "high" {
			reasoning = &struct {
				Effort ReasoningEffort `json:"effort"`
			}{Effort: ReasoningEffort(r)}
		}
	}

	rawBase := baseURL
	if rawBase == "" {
		rawBase = os.Getenv("OPENAI_BASE_URL")
	}
	if rawBase == "" {
		rawBase = os.Getenv("OPENAI_API_BASE")
	}
	if rawBase == "" {
		rawBase = DefaultOpenAIBaseURL
	}
	rawBase = strings.TrimRight(rawBase, "/")

	return ResolvedProviderRequest{
		Transport:      TransportChatCompletions,
		RequestedModel: requested,
		ResolvedModel:  resolvedModel,
		BaseURL:        rawBase,
		Reasoning:      reasoning,
	}
}
